safenode/src/network/mod.rs

```diff
@@ -0,0 +1,376 @@
+    swarm::{NetworkBehaviour, Swarm, SwarmBuilder, SwarmEvent},
+    PeerId,
+};
+use libp2p::{mdns, Multiaddr, Transport};
+use std::collections::{HashMap, HashSet};
+use std::{iter, time::Duration};
+use tracing::info;
+use xor_name::XorName;
+
+use self::chunk_codec::{ChunkRequest, ChunkResponse, ChunkStorageCodec, ChunkStorageProtocol};
+use self::command::CmdToSwarm;
+
+#[derive(Clone)]
+pub struct Client {
+    sender: mpsc::Sender<CmdToSwarm>,
+}
+
+impl Client {
+    //  Listen for incoming connections on the given address.
+    pub async fn start_listening(&mut self, addr: Multiaddr) -> Result<()> {
+        let (sender, receiver) = oneshot::channel();
+        self.sender
+            .send(CmdToSwarm::StartListening { addr, sender })
+            .await
+            .expect("Command receiver not to be dropped.");
+        receiver.await.expect("Sender not to be dropped.")
+    }
+
+    /// Dial the given peer at the given address.
+    pub async fn dial(&mut self, peer_id: PeerId, peer_addr: Multiaddr) -> Result<()> {
+        let (sender, receiver) = oneshot::channel();
+        self.sender
+            .send(CmdToSwarm::Dial {
+                peer_id,
+                peer_addr,
+                sender,
+            })
+            .await
+            .expect("Command receiver not to be dropped.");
+        receiver.await.expect("Sender not to be dropped.")
+    }
+
+    /// Advertise the local node as the provider of the given file on the DHT.
+    pub async fn store_chunk(&mut self, xor_name: XorName) {
+        let (sender, receiver) = oneshot::channel();
+        self.sender
+            .send(CmdToSwarm::StoreChunk { xor_name, sender })
+            .await
+            .expect("Command receiver not to be dropped.");
+        receiver.await.expect("Sender not to be dropped.");
+    }
+
+    /// Find the providers for the given file on the DHT.
+    pub async fn get_chunk_providers(&mut self, xor_name: XorName) -> HashSet<PeerId> {
+        let (sender, receiver) = oneshot::channel();
+        self.sender
+            .send(CmdToSwarm::GetChunkProviders { xor_name, sender })
+            .await
+            .expect("Command receiver not to be dropped.");
+        receiver.await.expect("Sender not to be dropped.")
+    }
+
+    /// Request the content of the given file from the given peer.
+    pub async fn request_chunk(&mut self, peer: PeerId, xor_name: XorName) -> Result<Vec<u8>> {
+        let (sender, receiver) = oneshot::channel();
+        self.sender
+            .send(CmdToSwarm::RequestChunk {
+                xor_name,
+                peer,
+                sender,
+            })
+            .await
+            .expect("Command receiver not to be dropped.");
+        receiver.await.expect("Sender not be dropped.")
+    }
+
+    /// Respond with the provided file content to the given request.
+    pub async fn respond_chunk(&mut self, file: Vec<u8>, channel: ResponseChannel<ChunkResponse>) {
```

Comment:
I guess this ha to be removed from the `Client`

Reply:
Renamed `Client` -> `NetworkApi` as the previous name was conflicting with our `Client` impl. The struct abstracts the underlying swarm.